default/python/AI/ProductionAI.py

```diff
@@ -96,15 +97,18 @@ def cur_best_military_design_rating():
     if current_turn in _best_military_design_rating_cache:
         return _best_military_design_rating_cache[current_turn]
     priority = PriorityType.PRODUCTION_MILITARY
-    if priority in _design_cache:
-        if _design_cache[priority] and _design_cache[priority][0]:
-            rating, pid, design_id, cost = _design_cache[priority][0]
-            _best_military_design_rating_cache[current_turn] = rating
-            return max(rating, 0.001)
-        else:
-            return 0.001
-    else:
-        return 0.001
+    if _design_cache.get(priority, None) and _design_cache[priority][0]:
+        # the rating provided by the ShipDesigner does not
+        # reflect the rating used in threat considerations
+        # but takes additional factors (such as cost) into
+        # account. Therefore, we want to calculate the actual
+        # rating of the design as provided by CombatRatingsAI.
+        _, _, _, _, stats = _design_cache[priority][0]
+        # TODO: Should this consider enemy stats?
+        rating = CombatRatingsAI.ShipCombatStats(stats=stats.convert_to_combat_stats()).get_rating()
```

Comment:
Having an argument for the enemy does sound best to me, but all the current uses would be against the standard enemy so I think the default should be to use the standard enemy rather than None.

There is only 1 current regular use, by PriorityAI.  The PriorityAI is trying to decide how much priorty to give to military ship production, comparing unmet threat (from a variety of sources) to the value returned by this function, to get an estimate of how many additional Military ships we feel an existing need for.  That could be made a more fine-grained analysis, but for now I would prefer it to use the standard enemy.

Also, in this new code it looks to me like you have the new method _combat_rating() *always* using the standard enemy when considering warship ratings, so it seems very consistent for us to have the standard enemy as default for cur_best_military_design_rating().

It is also currently optionally referenced in by MilitaryAI.cur_best_mil_ship_rating(), and that itself is used in a number of places; this function is getting the rating of our best existing ship (and optionally design), the ship rating is always gotten against the standard enemy and so if any of the places calling there wanted to also check designs they would need them evaluated against the standard enemy also.